### PR TITLE
Send switch state to the client

### PR DESCRIPTION
**Pull request by edg-l**
This PR is related to #3990

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test if it works standalone, system.c especially
- [ ] Considered possible null pointers and out of bounds array indexing
- [x] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by edg-l** [comment]
I'm doing this draft to know if something needs to be changed before doing more work on this.

Feel free to say anything

File: src/game/server/entities/character.cpp [hunk 1]
```diff
@@ -1246,6 +1246,44 @@ void CCharacter::Snap(int SnappingClient)
 
 	SnapCharacter(SnappingClient, ID);
 
+	if(Teams()->NeedsSwitchUpdate(Team()))
+	{
+		CNetObj_SwitchState *pSwitchState = static_cast<CNetObj_SwitchState *>(Server()->SnapNewItem(NETOBJTYPE_SWITCHSTATE, ID, sizeof(CNetObj_SwitchState)));
```
Comment (Chairn) [inline src/game/server/entities/character.cpp:1251]: You set pSwitchState then test if it is null, this is prone to segfault

File: src/game/server/entities/character.cpp [hunk 2]
```diff
@@ -1246,6 +1246,44 @@ void CCharacter::Snap(int SnappingClient)
...
+		pSwitchState->m_Status2 = 0;
+		pSwitchState->m_Status3 = 0;
+		pSwitchState->m_Status4 = 0;
+		pSwitchState->m_Status5 = 0;
+		pSwitchState->m_Status6 = 0;
+		pSwitchState->m_Status7 = 0;
+		pSwitchState->m_Status8 = 0;
+
+		if(pSwitchState)
+		{
```
Comment (Chairn) [inline src/game/server/entities/character.cpp:1262]: You set pSwitchState then test if it is null, this is prone to segfault

Comment (Learath2) [reply src/game/server/entities/character.cpp:1262]: I guess you mean the null check should happen immediately after the `SnapNewItem`, in which case I agree

Comment (Chairn) [reply src/game/server/entities/character.cpp:1262]: Yup, i mean you set to pSwitchState member then test if pSwitchState is null ;)

**Comment by trml** [comment]
I think the switch state has to be snapped every tick, so that it doesn't get dropped from the snapshot on the next tick/snap. (Not 100% sure, but it looked like that in the debug prints from gameclient.cpp)

**Comment by Learath2** [comment]
Yes, this kinda looks wrong. You should snap every tick, delta compression will take care of not updating all the time

**Comment by edg-l** [comment]
> Yes, this kinda looks wrong. You should snap every tick, delta compression will take care of not updating all the time

Oh i didnt know that, i guess i need to check more carefully how this snapshot system works

**Comment by edg-l** [comment]
This pr is about sending the switch data (not making the pred code, ill leave that to @trml ) I think it's done now, tell me if there is anything

**Comment by Learath2** [comment]
Since there is no good way to do any backwards/forwards compatibility in the protocol at this time I guess this is good to go.

bors r+

**Comment by fokkonaut** [comment]
Do we really need to send 64 instances of this? Wouldnt it be enough to only send one switchstate at a time, according to your own team?

**Comment by Learath2** [comment]
I'm guessing switch states for each team would be useful for predicting other tees.

NumSwitchers is just easier to send along instead of locally scanning the map and risking a buffer overflow e.g. when the client and server accidentally load different maps.

Both would be decent optimizations if delta compression wasn't a thing. Neither the switch state nor the number of switches would be sent with any regularity at all for this to matter. Maybe on a 128p server it could be an issue but not like that is a supported configuration.

We could send the switchstate per team instead if you are very concerned

**Comment by fokkonaut** [comment]
While the client loads the map, it has the m_NumSwitchers anyways. Its shared code. Its not an issue for 128p server, it would just make more sense to only send your own switch configuration, in order to predict tees from your team, tees that can collide with yourself.
Players from other teams dont have to be predicted, because you dont interact with them anyways.

**Pull request merged**